src/QsCompiler/Transformations/CallGraph/ConcreteCallGraphWalker.cs

```diff
@@ -245,7 +267,32 @@ public NamespaceWalker(SyntaxTreeTransformation<TransformationState> parent) : b
 
                 public override QsGeneratorDirective OnGeneratedImplementation(QsGeneratorDirective directive)
                 {
-                    throw new ArgumentException("Encountered unresolved generated specialization while constructing concrete call graph.");
+                    if (directive.IsSelfInverse)
+                    {
+                        if (this.SharedState.CurrentNode is null)
+                        {
+                            throw new ArgumentException("CurrentNode is expected to be non-null when processing self-adjoint specializations.");
+                        }
+
+                        if (this.SharedState.CurrentNode.Kind.IsQsAdjoint)
+                        {
+                            this.SharedState.AddSelfInverseDependency(this.SharedState.CurrentNode.CallableName, QsSpecializationKind.QsBody);
+                        }
+                        else if (this.SharedState.CurrentNode.Kind.IsQsControlledAdjoint)
+                        {
+                            this.SharedState.AddSelfInverseDependency(this.SharedState.CurrentNode.CallableName, QsSpecializationKind.QsControlled);
+                        }
+                        else
+                        {
+                            throw new ArgumentException("\"self\" can only be used on adjoint and controlled adjoint specializations.");
+                        }
+
+                        return directive;
+                    }
+                    else
+                    {
+                        throw new ArgumentException("Encountered unresolved generated specialization while constructing concrete call graph.");
```

Comment:
There should be a doc comment somewhere that needs to be updated to reflect that a "self" generation directive is ok.

Reply:
Updated the doc comment for the concrete call graph walker.